daemon/logger/amqp/amqp.go

```diff
@@ -0,0 +1,419 @@
+		logrus.Info("Enabling publish confirmation")
+		if err := c.Confirm(false); err != nil {
+			logrus.Errorf("Could not put channel into confirm mode - %v", err)
+			return nil, err
+		}
+
+		conf = c.NotifyPublish(make(chan amqp.Confirmation, 1))
+	}
+
+	err = c.ExchangeDeclare(currentBroker.Exchange, "direct", true, false, false, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not create exchange - %v", err)
+		return nil, err
+	}
+
+	_, err = c.QueueDeclare(currentBroker.Queue, true, false, false, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not create queue - %v", err)
+		return nil, err
+	}
+
+	err = c.QueueBind(currentBroker.Queue, currentBroker.RoutingKey, currentBroker.Exchange, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not bind queue to exchange - %v", err)
+		return nil, err
+	}
+
+	logrus.Info("Connection set up")
+	return &amqpConnection{
+		broker:     broker,
+		brokerURLs: brokerURLs,
+		conn:       conn,
+		c:          c,
+		conf:       conf,
+		err:        err,
+	}, nil
+}
+
+// If the connection fails at any point then close the current connection and
+// try to connect to the next broker in the list.
+func reconnect(s *amqpLogger) (err error) {
+	logrus.Warn("Unable to send message to AMQP broker")
+	logrus.Info("Attempting to reconnect")
+	s.Close()
+	// Move to the next broker in the list. If at the end of the
+	// list then go back to the start
+	if len(s.connection.brokerURLs) > s.connection.broker+1 {
+		s.connection.broker++
+	} else {
+		s.connection.broker = 0
+	}
+	connection, err := connect(s.ctx, s.connection.broker)
+	if err != nil {
+		logrus.Errorf("Could not reconnect: %v", err)
+		return err
+	}
+	logrus.Info("Reconnected")
+	s.connection = connection
+	return nil
+}
+
+// Take the log message and publish it to the currently connected broker
+func (s *amqpLogger) Log(msg *logger.Message) (err error) {
+	// Remove trailing and leading whitespace
+	short := bytes.TrimSpace([]byte(msg.Line))
+
+	if s.connection == nil {
+		err = reconnect(s)
+		if err != nil {
+			return err
+		}
+	}
+
+	currentBroker := s.connection.brokerURLs[s.connection.broker]
+
+	// If the message isn't empty then send it to the broker
+	if string(short) != "" {
+		m := amqpMessage{
+			Version:   "1",
+			Host:      s.fields.Hostname,
+			Message:   string(short),
+			Timestamp: time.Now(),
+			Path:      s.fields.ContainerID,
+			Tags:      s.fields,
+		}
+
+		messagejson, err := json.Marshal(m)
+		if err != nil {
+			logrus.Errorf("Could not serialise event - %v", err)
+			return err
+		}
+
+		amqpmsg := amqp.Publishing{
+			DeliveryMode: amqp.Persistent,
+			Timestamp:    time.Now(),
+			ContentType:  "application/json",
+			Body:         messagejson,
+		}
+
+		if currentBroker.Confirm == true && s.connection != nil {
+			defer confirmOne(s.connection.conf)
+		}
+
+		err = s.connection.c.Publish(currentBroker.Exchange, currentBroker.RoutingKey, false, false, amqpmsg)
+		if err != nil {
+			err = reconnect(s)
```

Comment:
Yes, I've thought about creating a local queue of messages in case there is a problem reconnecting but I've held off for now. With this log driver and a clustered AMQP broker set up I was aiming to avoid the need for a local store of log messages.